provider/sidecar.go

```diff
@@ -70,150 +89,171 @@ func (provider *Sidecar) Provide(configurationChan chan<- types.ConfigMessage, p
 					return
 				case event := <-watcher.Events:
 					if strings.Contains(event.Name, file.Name()) {
-						log.Debug("Sidecar Frontend File event:", event)
-						states, errState := provider.fetchState()
-						if errState != nil {
-							log.Errorln("Error reloading Sidecar config", errState)
+						log.Debug("Sidecar config file event: ", event)
+						err = provider.reloadConfig()
+						if err != nil {
+							log.Error(err)
 						}
-						provider.loadSidecarConfig(states.ByService())
 					}
 				case errWatcher := <-watcher.Errors:
-					log.Errorln("Watcher event error", errWatcher)
+					log.Error("Watcher event error: ", errWatcher)
 				}
 			}
 		})
 		err = watcher.Add(filepath.Dir(file.Name()))
 		if err != nil {
-			log.Error("Error adding file watcher", err)
+			log.Error("Error adding file watcher: ", err)
 			return err
 		}
+	} else {
+		err := provider.reloadConfig()
+		if err != nil {
+			log.Error(err)
+		}
 	}
+
+	return nil
+}
+
+func (provider *Sidecar) reloadConfig() error {
 	states, err := provider.fetchState()
 	if err != nil {
-		log.Fatalln("Error reloading Sidecar config", err)
+		return fmt.Errorf("Error fetching Sidecar state: %s", err)
 	}
-	err = provider.loadSidecarConfig(states.ByService())
+
+	err = provider.loadConfig(states.ByService())
 	if err != nil {
-		return err
+		return fmt.Errorf("Error loading Sidecar config: %s", err)
 	}
+
 	return nil
 }
 
-func (provider *Sidecar) constructConfig(sidecarStates map[string][]*service.Service) (*types.Configuration, error) {
-	log.Infoln("loading sidecar config")
-	sidecarConfig := types.Configuration{Backends: provider.makeBackends(sidecarStates)}
-	var err error
-	sidecarConfig.Frontends, err = provider.makeFrontend()
-	if err != nil {
-		return nil, err
+func (provider *Sidecar) loadConfig(sidecarStates map[string][]*service.Service) error {
+	log.Info("Loading sidecar config...")
+
+	config := &types.Configuration{
+		Backends: make(map[string]*types.Backend),
 	}
-	return &sidecarConfig, nil
-}
 
-func (provider *Sidecar) loadSidecarConfig(sidecarStates map[string][]*service.Service) error {
-	conf, err := provider.constructConfig(sidecarStates)
-	if err != nil {
+	if _, err := toml.DecodeFile(provider.Filename, config); err != nil {
 		return err
 	}
+
+	// Create backends from Sidecar state data
+	for serviceName, services := range sidecarStates {
```

Comment:
Here we could just pass the whole state in and call `EachService` on it rather than first generating the map and then passing it here and then ranging over it. I know this is different from the original code, too. But if we were going to redo it, that's what I'd recommend over this.